File: sdk/typescript/src/clients/pulls.ts
/**
 * Pull requests resource client.
 *
 * Design Reference: DR-5
 * Requirements: 9.1, 9.2, 9.3, 9.4, 9.5
 */

import type { HTTPTransport } from '../transport.js';
import type { DiffStats, MergeResult, PullRequest } from '../types/index.js';

/**
 * Parse ISO date string to Date object.
 */
function parseDate(dateStr: string): Date {
  return new Date(dateStr.replace(/Z$/, '+00:00'));
}

/**
 * Parse pull request data from API response.
 */
function parsePullRequest(data: Record<string, unknown>): PullRequest {
  const diffStatsData = (data.diffStats ?? {}) as Record<string, unknown>;
  const diffStats: DiffStats = {
    filesChanged: (diffStatsData.filesChanged as number) ?? 0,
    insertions: (diffStatsData.insertions as number) ?? 0,
    deletions: (diffStatsData.deletions as number) ?? 0,
  };

  let mergedAt: Date | null = null;
  if (data.mergedAt) {
    mergedAt = parseDate(data.mergedAt as string);
  }

  return {
    prId: data.prId as string,
    repoId: data.repoId as string,
    authorId: data.authorId as string,
    sourceBranch: data.sourceBranch as string,
    targetBranch: data.targetBranch as string,
    title: data.title as string,
    description: (data.description as string | null) ?? null,
    status: data.status as 'open' | 'merged' | 'closed',
    ciStatus: (data.ciStatus as 'pending' | 'running' | 'passed' | 'failed') ?? 'pending',
    diffStats,
    mergeable: (data.mergeable as boolean) ?? false,
    isApproved: (data.isApproved as boolean) ?? false,
    reviewCount: (data.reviewCount as number) ?? 0,
    createdAt: parseDate(data.createdAt as string),
    mergedAt,
  };
}

/**
 * Client for pull request operations.
 *
 * Design Reference: DR-5
 */
export class PullsClient {
  private transport: HTTPTransport;

  /**
   * Initialize the pulls client.
   *
   * @param transport - HTTP transport for making requests
   */
  constructor(transport: HTTPTransport) {
    this.transport = transport;
  }

  /**
   * Create a pull request.
   *
   * @param repoId - The repository identifier
   * @param sourceBranch - Branch containing changes
   * @param targetBranch - Branch to merge into
   * @param title - Pull request title
   * @param description - Optional pull request description
   * @returns PullRequest with prId, ciStatus, diffStats, mergeable status
   * @throws AuthenticationError if signature is invalid
   * @throws ValidationError if branch not found
   * @throws NotFoundError if repository not found
   *
   * Requirements: 9.1, 9.2
   */
  async create(
    repoId: string,
    sourceBranch: string,
    targetBranch: string,
    title: string,
    description?: string | null
  ): Promise<PullRequest> {
    const body: Record<string, unknown> = {
      repoId,
      sourceBranch,
      targetBranch,
      title,
      description: description ?? null,
    };

    const response = await this.transport.signedRequest(
      'POST',
      `/v1/repos/${repoId}/pulls`,
      'pr_create',
      body
    );

    const data = (response.data ?? {}) as Record<string, unknown>;
    return parsePullRequest(data);
  }

  /**
   * Get pull request information.
   *
   * @param repoId - The repository identifier
   * @param prId - The pull request identifier
   * @returns PullRequest with full details
   * @throws NotFoundError if pull request not found
   *
   * Requirements: 9.2
   */
  async get(repoId: string, prId: string): Promise<PullRequest> {
    const response = await this.transport.unsignedRequest(
      'GET',
      `/v1/repos/${repoId}/pulls/${prId}`
    );

    const data = (response.data ?? {}) as Record<string, unknown>;
    return parsePullRequest(data);
  }

  /**
   * List pull requests.
   *
   * @param repoId - The repository identifier
   * @param status - Optional filter by status ("open", "merged", "closed")
   * @param authorId - Optional filter by author
   * @returns List of PullRequest objects
   *
   * Requirements: 9.2
   */
  async list(
    repoId: string,
    status?: 'open' | 'merged' | 'closed',
    authorId?: string
  ): Promise<PullRequest[]> {
    const params: Record<string, string> = {};
    if (status) {
      params.status = status;
    }
    if (authorId) {
      params.authorId = authorId;
    }

    const response = await this.transport.unsignedRequest(
      'GET',
      `/v1/repos/${repoId}/pulls`,
      Object.keys(params).length > 0 ? params : undefined
    );

    const data = (response.data ?? {}) as Record<string, unknown>;
    const pulls = (data.pulls ?? []) as Record<string, unknown>[];
    return pulls.map(parsePullRequest);
  }

  /**
   * Merge a pull request.
   *
   * @param repoId - The repository identifier
   * @param prId - The pull request identifier
   * @param mergeStrategy - "merge", "squash", or "rebase" (default: "merge")
   * @returns MergeResult with mergeCommitOid
   * @throws AuthenticationError if signature is invalid
   * @throws ValidationError if PR not approved or CI not passed
   * @throws ConflictError if merge conflicts or already merged
   * @throws NotFoundError if pull request not found
   *
   * Requirements: 9.4, 9.5
   */
  async merge(
    repoId: string,
    prId: string,
    mergeStrategy: 'merge' | 'squash' | 'rebase' = 'merge'
  ): Promise<MergeResult> {
    const response = await this.transport.signedRequest(
      'POST',
      `/v1/repos/${repoId}/pulls/${prId}/merge`,
      'pr_merge',
      {
        repoId,
        prId,
        mergeStrategy,
      }
    );

    const data = (response.data ?? {}) as Record<string, unknown>;
    return {
      prId: data.prId as string,
      repoId: data.repoId as string,
      mergeStrategy: data.mergeStrategy as string,
      mergedAt: parseDate(data.mergedAt as string),
      mergeCommitOid: data.mergeCommitOid as string,
    };
  }
}

File: sdk/rust/src/clients/pulls.rs
//! Pull requests resource client.
//!
//! Design Reference: DR-5
//! Requirements: 9.1, 9.2, 9.4, 9.5

use std::collections::HashMap;
use std::sync::Arc;

use serde_json::Value;

use crate::error::Error;
use crate::transport::HttpTransport;
use crate::types::{MergeResult, PullRequest};

/// Client for pull request operations.
pub struct PullsClient {
    transport: Arc<HttpTransport>,
}

impl PullsClient {
    /// Create a new pulls client.
    pub fn new(transport: Arc<HttpTransport>) -> Self {
        Self { transport }
    }

    /// Create a pull request.
    ///
    /// # Arguments
    ///
    /// * `repo_id` - The repository identifier
    /// * `source_branch` - Source branch name
    /// * `target_branch` - Target branch name
    /// * `title` - Pull request title
    /// * `description` - Optional pull request description
    ///
    /// # Returns
    ///
    /// PullRequest object with pr_id, ci_status, diff_stats, etc.
    ///
    /// # Errors
    ///
    /// Returns an error if creation fails.
    ///
    /// Requirements: 9.1, 9.2
    pub async fn create(
        &self,
        repo_id: &str,
        source_branch: &str,
        target_branch: &str,
        title: &str,
        description: Option<&str>,
    ) -> Result<PullRequest, Error> {
        let mut body: HashMap<String, Value> = HashMap::new();
        body.insert("repoId".to_string(), Value::String(repo_id.to_string()));
        body.insert(
            "sourceBranch".to_string(),
            Value::String(source_branch.to_string()),
        );
        body.insert(
            "targetBranch".to_string(),
            Value::String(target_branch.to_string()),
        );
        body.insert("title".to_string(), Value::String(title.to_string()));
        body.insert(
            "description".to_string(),
            description.map_or(Value::Null, |d| Value::String(d.to_string())),
        );

        let response: Value = self
            .transport
            .signed_request(
                "POST",
                &format!("/v1/repos/{repo_id}/pulls"),
                "pr_create",
                body,
            )
            .await?;

        let data = response
            .get("data")
            .ok_or_else(|| Error::Http("Missing data in response".to_string()))?;

        serde_json::from_value(data.clone()).map_err(Error::from)
    }

    /// Get pull request information.
    ///
    /// # Arguments
    ///
    /// * `repo_id` - The repository identifier
    /// * `pr_id` - The pull request identifier
    ///
    /// # Returns
    ///
    /// PullRequest object
    ///
    /// # Errors
    ///
    /// Returns an error if the PR is not found.
    pub async fn get(&self, repo_id: &str, pr_id: &str) -> Result<PullRequest, Error> {
        let response: Value = self
            .transport
            .unsigned_request::<Value>(
                "GET",
                &format!("/v1/repos/{repo_id}/pulls/{pr_id}"),
                None,
                None::<&()>,
            )
            .await?;

        let data = response
            .get("data")
            .ok_or_else(|| Error::Http("Missing data in response".to_string()))?;

        serde_json::from_value(data.clone()).map_err(Error::from)
    }

    /// List pull requests.
    ///
    /// # Arguments
    ///
    /// * `repo_id` - The repository identifier
    /// * `status` - Optional status filter ("open", "merged", "closed")
    /// * `author_id` - Optional author filter
    ///
    /// # Returns
    ///
    /// List of PullRequest objects
    ///
    /// # Errors
    ///
    /// Returns an error if the request fails.
    pub async fn list(
        &self,
        repo_id: &str,
        status: Option<&str>,
        author_id: Option<&str>,
    ) -> Result<Vec<PullRequest>, Error> {
        let mut params: Vec<(&str, &str)> = Vec::new();
        if let Some(s) = status {
            params.push(("status", s));
        }
        if let Some(a) = author_id {
            params.push(("authorId", a));
        }

        let params_ref = if params.is_empty() {
            None
        } else {
            Some(params.as_slice())
        };

        let response: Value = self
            .transport
            .unsigned_request::<Value>(
                "GET",
                &format!("/v1/repos/{repo_id}/pulls"),
                params_ref,
                None::<&()>,
            )
            .await?;

        let data = response
            .get("data")
            .ok_or_else(|| Error::Http("Missing data in response".to_string()))?;

        let pulls = data
            .get("pulls")
            .ok_or_else(|| Error::Http("Missing pulls in response".to_string()))?;

        serde_json::from_value(pulls.clone()).map_err(Error::from)
    }

    /// Merge a pull request.
    ///
    /// # Arguments
    ///
    /// * `repo_id` - The repository identifier
    /// * `pr_id` - The pull request identifier
    /// * `merge_strategy` - Merge strategy: "merge", "squash", or "rebase"
    ///
    /// # Returns
    ///
    /// MergeResult with merge_commit_oid
    ///
    /// # Errors
    ///
    /// Returns an error if merge fails.
    ///
    /// Requirements: 9.4, 9.5
    pub async fn merge(
        &self,
        repo_id: &str,
        pr_id: &str,
        merge_strategy: Option<&str>,
    ) -> Result<MergeResult, Error> {
        let mut body: HashMap<String, Value> = HashMap::new();
        body.insert("repoId".to_string(), Value::String(repo_id.to_string()));
        body.insert("prId".to_string(), Value::String(pr_id.to_string()));
        body.insert(
            "mergeStrategy".to_string(),
            Value::String(merge_strategy.unwrap_or("merge").to_string()),
        );

        let response: Value = self
            .transport
            .signed_request(
                "POST",
                &format!("/v1/repos/{repo_id}/pulls/{pr_id}/merge"),
                "pr_merge",
                body,
            )
            .await?;

        let data = response
            .get("data")
            .ok_or_else(|| Error::Http("Missing data in response".to_string()))?;

        serde_json::from_value(data.clone()).map_err(Error::from)
    }
}
